src/struct_metadata.rs

```diff
@@ -0,0 +1,166 @@
+    /// (De)serializing a struct as a map will be approximately 40% slower than as a sequence.
+    access_type: AccessType,
+}
+
+impl StructMetadata {
+    // FIXME: perhaps it should not be public? But it is required for mocks/provide.
+    pub fn new<T: Row>(columns: Vec<Column>) -> Self {
+        let struct_name = T::NAME;
+        let struct_fields = T::COLUMN_NAMES;
+        if columns.len() != struct_fields.len() {
+            panic!(
+                "While processing struct {}: database schema has {} columns, \
+                but the struct definition has {} fields.\
+                \n#### All struct fields:\n{}\n#### All schema columns:\n{}",
+                struct_name,
+                columns.len(),
+                struct_fields.len(),
+                join_panic_schema_hint(struct_fields),
+                join_panic_schema_hint(&columns),
+            );
+        }
+        let mut mapping = Vec::with_capacity(struct_fields.len());
+        let mut expected_index = 0;
+        let mut should_use_map = false;
+        for col in &columns {
+            if let Some(index) = struct_fields.iter().position(|field| col.name == *field) {
+                if index != expected_index {
+                    should_use_map = true
+                }
+                expected_index += 1;
+                mapping.push(index);
+            } else {
+                panic!(
+                    "While processing struct {}: database schema has a column {} \
+                    that was not found in the struct definition.\
+                    \n#### All struct fields:\n{}\n#### All schema columns:\n{}",
+                    struct_name,
+                    col,
+                    join_panic_schema_hint(struct_fields),
+                    join_panic_schema_hint(&columns),
+                );
+            }
+        }
+        Self {
+            columns,
+            struct_name,
+            struct_fields,
+            row_type: T::TYPE,
+            access_type: if should_use_map {
+                AccessType::WithMapAccess(mapping)
+            } else {
+                AccessType::WithSeqAccess
+            },
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) fn is_struct(&self) -> bool {
+        matches!(self.row_type, RowType::Struct)
+    }
+
+    #[inline(always)]
+    pub(crate) fn get_schema_index(&self, struct_idx: usize) -> usize {
+        match &self.access_type {
+            AccessType::WithMapAccess(mapping) => {
+                if struct_idx < mapping.len() {
+                    mapping[struct_idx]
+                } else {
+                    panic!(
+                        "Struct {} has more fields than columns in the database schema",
+                        self.struct_name
+                    )
+                }
+            }
+            AccessType::WithSeqAccess => struct_idx, // should be unreachable
+        }
+    }
+}
+
+pub(crate) async fn get_struct_metadata<T: Row>(
```

Comment:
Is it used?

Reply:
should be used when `insert` is implemented, see the dead code comment